### PR TITLE
Links: Assure base url when single stat, panel and data links are built

### LINKED ISSUE #21839: Inconsistent behavior with relative URLs in SingleStat panel link
**What happened**: My Grafana instance is exposed in a subpath, *http://server/grafana*. When setting a Panel link in a SingleStat panel with a URL like */grafana/d/mydashboard*, the link behavior is inconsistent:

-  The link that appears in the top-left corner popup points to *http://server/grafana/d/mydashboard*
- Clicking on the panel redirects to *http://server/grafana/grafana/d/mydashboard* (it adds the base URL configured in Grafana) but only when 'Open in new tab' is not selected. If I select the new tab option, the URL in the new tab is *http://server/grafana/d/mydashboard*.

**What you expected to happen**: Links should behave the same way in all of those cases.

**How to reproduce it (as minimally and precisely as possible)**:

1. Start a Grafana instance in a subpath of a server
2. Add a new dashboard with a SingleStat panel.
3. Add a Panel link in the panel that starts with "/".

**Anything else we need to know?**: It seems that the click() event handler for the SingleStat panel manages URLs differently, with window.open() if it wants to open in a new tab or an absolute URL, and with $location.url for other cases. But that code wasn't changed recently so it seems that it is a change of behavior of the $location service.

**Environment**:
- Grafana version: Version 6.6.0 (commit: 5c11bbdfb4, branch: HEAD)
- Data source type & version: N/A
- OS Grafana is installed on: Ubuntu 18.04 LTS
- User OS & Browser: macOS, Firefox 73.0b9 and Chrome 79.0
- Grafana plugins: N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by dprokop**
Fixes #21839

Fixes regression caused by https://github.com/grafana/grafana/pull/20161

### CODE REVIEW

File: public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderCorner.tsx [hunk 1]
```diff
@@ -56,9 +58,14 @@ export class PanelHeaderCorner extends Component<Props> {
         {links && links.length > 0 && (
           <ul className="panel-info-corner-links">
             {links.map((link, idx) => {
+              console.log('corner', `${getConfig().appSubUrl}${locationUtil.stripBaseFromUrl(link.href)}`);
```
Comment (dprokop) [inline public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderCorner.tsx]:
```suggestion
```

File: public/app/features/panel/panellinks/link_srv.ts [hunk 2]
```diff
@@ -215,8 +216,17 @@ export class LinkSrv implements LinkService {
   /** @ngInject */
   constructor(private templateSrv: TemplateSrv, private timeSrv: TimeSrv) {}
 
+  assureBaseUrl(url: string) {
+    console.log('assuring', url);
```
Comment (dprokop) [inline public/app/features/panel/panellinks/link_srv.ts]:
```suggestion
```

File: public/app/features/panel/panellinks/link_srv.ts [hunk 3]
```diff
@@ -215,8 +216,17 @@ export class LinkSrv implements LinkService {
...
+    if (url.startsWith('/')) {
+      return `${getConfig().appSubUrl}${locationUtil.stripBaseFromUrl(url)}`;
+    }
+    return url;
+  }
+
   getLinkUrl(link: any) {
-    const url = this.templateSrv.replace(link.url || '');
+    const url = this.assureBaseUrl(this.templateSrv.replace(link.url || ''));
+
```
Comment (dprokop) [inline public/app/features/panel/panellinks/link_srv.ts]:
```suggestion
```

File: public/app/features/panel/panellinks/link_srv.ts [hunk 4]
```diff
@@ -286,7 +296,9 @@ export class LinkSrv implements LinkService {
         value: variablesQuery,
       },
     });
+    console.log(info);
```
Comment (dprokop) [inline public/app/features/panel/panellinks/link_srv.ts]:
```suggestion
```

File: public/app/features/panel/panellinks/link_srv.ts [hunk 5]
```diff
@@ -286,7 +296,9 @@ export class LinkSrv implements LinkService {
...
     info.href = getConfig().disableSanitizeHtml ? info.href : sanitizeUrl(info.href);
+    console.log(info);
```
Comment (dprokop) [inline public/app/features/panel/panellinks/link_srv.ts]:
```suggestion
```

File: public/app/plugins/panel/singlestat/module.ts [hunk 6]
```diff
@@ -653,7 +654,7 @@ class SingleStatCtrl extends MetricsPanelCtrl {
           window.location.href = linkInfo.href;
         } else {
           $timeout(() => {
-            $location.url(linkInfo.href);
+            $location.url(locationUtil.stripBaseFromUrl(linkInfo.href));
```
Comment (dprokop) [inline public/app/plugins/panel/singlestat/module.ts]: This is weird, I know. But somehow $location service has base url set, hence the url has double base urls in it. @torkelo do you know why $location behaves like this?

Comment (torkelo) [reply public/app/plugins/panel/singlestat/module.ts]: No, cant remember a reason

Comment (dprokop) [reply public/app/plugins/panel/singlestat/module.ts]: was trying to find some location service base config, but no trace. Maybe this is just how it is 🤷‍♀

File: public/app/features/panel/panellinks/link_srv.ts [hunk 7]
```diff
@@ -215,8 +216,15 @@ export class LinkSrv implements LinkService {
   /** @ngInject */
   constructor(private templateSrv: TemplateSrv, private timeSrv: TimeSrv) {}
 
+  assureBaseUrl(url: string) {
```
Comment (hugohaggmark) [inline public/app/features/panel/panellinks/link_srv.ts]: Nit: could assureBaseUrl be part of locationUtil?

File: public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderCorner.tsx [hunk 8]
```diff
@@ -58,7 +60,11 @@ export class PanelHeaderCorner extends Component<Props> {
             {links.map((link, idx) => {
               return (
                 <li key={idx}>
-                  <a className="panel-info-corner-links__item" href={link.href} target={link.target}>
+                  <a
+                    className="panel-info-corner-links__item"
+                    href={`${getConfig().appSubUrl}${locationUtil.stripBaseFromUrl(link.href)}`}
```
Comment (hugohaggmark) [inline public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderCorner.tsx]: Just a thought are all urls in this PR are sanitized already?

**Review by hugohaggmark — APPROVED** [review]
Looks good to me, but haven't tested it. Left just small nits.

**Comment by dprokop** [comment]
@torkelo do you mind taking a look at this in free time?

File: public/app/core/utils/location_util.ts [hunk 9]
```diff
@@ -1,12 +1,19 @@
...
 
-export default { stripBaseFromUrl };
+export const assureBaseUrl = (url: string) => {
+  if (url.startsWith('/')) {
+    return `${getConfig().appSubUrl}${stripBaseFromUrl(url)}`;
+  }
+  return url;
+};
+
+export default { stripBaseFromUrl, assureBaseUrl };
```
Comment (hugohaggmark) [inline public/app/core/utils/location_util.ts]: Nice 🥇

**Review by torkelo — APPROVED** [review]
Looks good!

**Pull request merged**